Refactor style logic to avoid direct access to the node data during the cascade

bholley opened the pull request:
The new restyle architecture doesn't store these things in consistent places, so we need a more abstract API.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13656)

<!-- Reviewable:end -->

bholley committed:
Refactor cascade to limit and clarify the mutability of the old style.

The current semantics make it very difficult to figure out when the existing
style is mutated and what code depends on it. This is problematic for the
new incremental restyle architecture, where the cascade logic no longer has
direct access to |data|.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/layout_wrapper.rs, components/script_layout_interface/lib.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style, layout, and script code, but no tests are modified. Please consider adding a test!

bholley commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0431ed6 with merge af10369...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/101)

highfive commented:
▶ TIMEOUT [expected FAIL] /css21_dev/html4/abspos-replaced-width-margin-000.htm
      │ 
      │ VMware, Inc.
      │ Gallium 0.4 on softpipe
      │ 3.3 (Core Profile) Mesa 12.0.1
      └ Shutting down the Constellation after generating an output file or exit flag specified

bholley committed:
Remove borrow_data and mutate_data from TNode.

The new restyle architecture doesn't store these things in consistent
places, so we need a more abstract API.

bholley commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7bf63d9 with merge 0b3cf7e...

bholley commented:
r? @emilio

bors-servo commented:
:broken_heart: Test failed - [windows-dev](http://build.servo.org/builders/windows-dev/builds/1377)

bholley commented:
ProtocolError: ("Connection broken: error(104, 'Connection reset by peer')", error(104, 'Connection reset by peer'))

@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 7bf63d9 with merge 4315172...

bholley commented:
@bors-servo try

bholley commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 46b2787 with merge 435d704...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3420), [arm64](http://build.servo.org/builders/arm64/builds/3432), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3506), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/127), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/127), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3497), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3478), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/178), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/177), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1385)

emilio commented on components/script/layout_wrapper.rs:
```diff
@@ -234,12 +242,23 @@ impl<'ln> TNode for ServoLayoutNode<'ln> {
+    }
+
+    fn set_pseudo_styles(&self, styles: PseudoStyles) {
+        self.mutate_data().unwrap().per_pseudo = styles;
```
`debug_assert!` that the current pseudo styles is empty?

I don't know if it'd be more efficient to make this something like an `Option`, then use `Option::take`. It'd avoid the construction of the hashmap on `take_pseudo_styles`.

Also, probably you want to open an issue to make the pseudo styles a vector (or a `SmallVec<[2; T]> / AutoTArray<2, T>` and not a hashmap.

bholley replied:
> debug_assert! that the current pseudo styles is empty?

Done.

I considered the Option thing, but I think it makes the semantics more confusing, and empty hashmap construction is cheap (there's no allocation). So I don't think it's worth it.

Converting the HashMap to a vec is on my TODO list (in the googledoc), but I'll do it once the architecture settles more. I'll have to change this pseudo stuff around a bit anyway to handle the fact that we may not have the hashmap in the gecko case.

emilio commented on components/style/gecko/wrapper.rs:
```diff
@@ -319,14 +333,23 @@ impl<'ln> TNode for GeckoNode<'ln> {
-    #[inline(always)]
-    fn borrow_data(&self) -> Option<AtomicRef<PersistentStyleData>> {
-        self.get_node_data().as_ref().map(|d| d.0.borrow())
+    fn get_existing_style(&self) -> Option<Arc<ComputedValues>> {
```
Note that this is also subtly changing the behavior of most of the other code that used something like `borrow_data().style.as_ref.unwrap()`. Probably you were also aware of it.

bholley replied:
You mean in terms of the extra refcount bump? I'm aware of that, but the borrow is pretty hard to preserve in the case where the computed style hangs off the nsIFrame. That's the reason for this refactoring.

If there are other behavior changes I'm not seeing please let me know.

emilio commented on components/style/matching.rs:
```diff
@@ -486,6 +484,14 @@ pub enum StyleSharingResult<ConcreteRestyleDamage: TRestyleDamage> {
+    shareable: bool,
+    cacheable: bool,
+    animate: bool,
+}
```
I think it'd be cleaner to use boolean enum for each one of those instead of a struct holding the three bools. Also, note that `cacheable` wasn't given as parameter to `cascade_node_pseudo_element`. Is that intentional, or just a consequence of it? It's probably ok in any case, I guess.

emilio commented on components/style/matching.rs:
```diff
@@ -724,14 +721,13 @@ pub trait ElementMatchMethods : TElement {
                     // replaced content, or similar stuff!
                     let damage =
-                        match node.existing_style_for_restyle_damage((*style).as_ref(), None) {
+                        match node.existing_style_for_restyle_damage(node.get_existing_style().as_ref(), None) {
```
(to elaborate on my previous comment, you're, for example, creating a spurious refcount increment/decrement here which wasn't needed before)

emilio commented on components/style/matching.rs:
```diff
@@ -907,23 +900,38 @@ pub trait MatchMethods : TNode {
+            let mut old_style = self.get_existing_style();
+            let cacheable = match old_style {
+                None => true,
+                Some(ref mut old) => {
```
Hmm... I see the `cacheable` thing now. It's kind of unfortunate we're moving this out of `cascade_node_pseudo_element`, which is conceptually "everything you need to do to get an element style. Is there an easy way to move it back in? Haven't thought too hard about it, probably it's not a big deal if not.

emilio commented on components/style/traversal.rs:
```diff
@@ -229,15 +229,6 @@ fn ensure_node_styled_internal<'a, N, C>(node: N,
-    //
-    // Ideally we'd have an initialize_data() or something similar but just for
-    // style data.
-    debug_assert!(node.borrow_data().is_some(),
```
Hmm... I guess there isn't an easy way to restore the equivalent of this assertion (which is better than "tried to unwrap None").

Can you keep the comment somewhere at least, saying that this expects the node data to be initialized?

bholley replied:
I've added a comment. I didn't sweat it too much before because the invariants here are all going to change with the new architecture anyway.

emilio reviewed (commented):
The changes look ok to me, modulo my comments/questions.

emilio commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 710f951 has been approved by `emilio`

<!-- @bors-servo r=emilio 710f9519e793e8229245249d7b19590a018a89a2 -->

bors-servo commented:
:hourglass: Testing commit 710f951 with merge 661b38a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/139)

bholley commented:
Doh, put the assert requested above in the wrong method.

@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit b4c0475 has been approved by `emilio`

<!-- @bors-servo r=emilio b4c0475ec18d11dbeedd53996147692efd164021 -->

bors-servo commented:
:hourglass: Testing commit b4c0475 with merge e82f046...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/192)

bholley commented:
@bors-servo retry #13662 p=1

bors-servo commented:
:hourglass: Testing commit b4c0475 with merge ff48ca1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/145)

highfive commented:
▶ Unexpected subtest result in /_mozilla/mozilla/promise.html:
      │ FAIL [expected PASS] Native promise from async callback can be resolved
      │   → assert_greater_than_equal: expected a number greater than or equal to 100 but got 99
      │ 
      └ @http://web-platform.test:8000/_mozilla/mozilla/promise.html:58:11

KiChjang commented:
@bors-servo retry
* #13684

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3438), [arm64](http://build.servo.org/builders/arm64/builds/3450), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3523), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/145), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3515), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3496), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/196), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/195), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1403) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/145)...

notriddle commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit bfbbef6 has been approved by `emilio`

<!-- @bors-servo r=emilio bfbbef6ecdfd8bdd0a58bd1a0c7ed95c182b98a7 -->

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #13679

bors-servo commented:
:pushpin: Commit bfbbef6 has been approved by `emilio`

<!-- @bors-servo r=emilio bfbbef6ecdfd8bdd0a58bd1a0c7ed95c182b98a7 -->

bors-servo commented:
:hourglass: Testing commit bfbbef6 with merge abcc4ae...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/199)

bholley commented:
@bors-servo retry ##13662

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3441), [arm64](http://build.servo.org/builders/arm64/builds/3453), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3527), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/148), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/148), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3518), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3499), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/199), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1406) are reusable. Rebuilding only [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/199)...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3441), [arm64](http://build.servo.org/builders/arm64/builds/3453), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3527), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/148), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/148), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3518), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3499), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/199), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/201), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1406)
